NServiceBus.Extensions.DependencyInjection documentation (#4691)
Particular/docs.particular.net: nservicebus/dependency-injection/nservicebus-dependencyinjection.md

---
title: NServiceBus.Extensions.DependencyInjection
summary: Provides integration with the Microsoft.Extensions.DependencyInjection abstraction.
reviewed: 2020-02-17
component: Extensions.DependencyInjection
tags:
 - Dependency Injection
related:
 - samples/dependency-injection/extensions.dependencyinjection
---

The `NServiceBus.Extensions.DependencyInjection` package provides integration with the `Microsoft.Extensions.DependencyInjection` dependency injection abstraction.

## Usage with ServiceCollection

The following snippet shows how to configure NServiceBus to use Microsoft's built-in dependency injection container:

snippet: usecontainer-servicecollection

## Usage with third party containers

> NServiceBus can also be configured to work with any third party dependency injection which implements the `Microsoft.Extensions.DependencyInjection` abstraction. To use a third party dependency injection container, pass the specfic container's `IServiceProviderFactory` to the `UseContainer` configuration method. The following snippet shows this approach, using Autofac (`Autofac.Extensions.DependencyInjection`) as an example:

Reviewer: ```suggestion
NServiceBus can also be configured to work with any third party dependency injection container which implements the `Microsoft.Extensions.DependencyInjection` abstraction. To use a third-party dependency injection container, pass the specific container's `IServiceProviderFactory` to the `UseContainer` configuration method. The following snippet shows this approach, using Autofac (`Autofac.Extensions.DependencyInjection`) as an example:
```

Author: I like those suggestions, but you're using third-party and third party, should align those cases?

snippet: usecontainer-thirdparty

## Configuring the container

`UseContainer` provides a settings class which gives advanced configuration options:

### IServiceCollection access

> Via the settings, the underlying `IServiceCollection` can be accessed to add additional service registrations.

Reviewer: ```suggestion
The settings provide access to the underlying `IServiceCollection` that can be used to add additional service registrations.
```

snippet: settings-servicecollection

### ContainerBuilder

> Native container APIs can be accessed by the `ConfigureContainer` API.

Reviewer: ```suggestion
Third-party container-native APIs can be accessed by with `ConfigureContainer`.
```

snippet: settings-configurecontainer

## Property Injection

> The `NServiceBus.Extensions.DependencyInjection` package does not support property injection out of the box. To enable property injection, please refer to the configured container's documentation.

Reviewer: ```suggestion
The `NServiceBus.Extensions.DependencyInjection` package does **not** support property injection out of the box. To enable property injection, refer to the configured container's documentation.
```
